### ISSUE TITLE
Provide a way for users to cutomize `derive`s on custom types

### ISSUE THREAD

**Issue by LouisGariepy** [LouisGariepy/cornucopia#25]
I don't want to presuppose too much about how the generated code will be used, thus, so far, I've refrained from adding any superfluous derives to the generated types.

This might be limiting to some users wishing to add `derive`s to their types (like `Serialize` or `Clone`). I'm thinking of flags like `--custom-struct-derive`, `--custom-enum-derive` and `--custom-derive` which could be used to add the `derive` only to `struct`s, `enum`s, or both.

This will require a bit more thought though, so if you're reading this thread because you want this feature, feel free to ping me and explain how you'd like this feature to work in practice.

**Comment by Virgiel** [comment]
I do not know if [SQLDelight](https://cashapp.github.io/sqldelight/) was an inspiration, but they also have the same idea of being SQL first. I do not think we can make an API as ergonomic as theirs as Kotlin is less rigid than Rust. To map query rows to structures they use [projections](https://cashapp.github.io/sqldelight/android_sqlite/custom_projections/), we could do something similar same with stream mapping.

**Comment by LouisGariepy** [comment]
> I do not know if [SQLDelight](https://cashapp.github.io/sqldelight/) was an inspiration...

I started this project thinking I had had a pretty novel idea :laughing: . Since starting though, I've been made aware of a number of similar projects and I've been looking into incorporating their most interesting features in Cornucopia.

>I do not think we can make an API as ergonomic as theirs as Kotlin is less rigid than Rust. To map query rows to structures they use [projections](https://cashapp.github.io/sqldelight/android_sqlite/custom_projections/), we could do something similar same with stream mapping.

Unfortunately, I don't really understand what you're suggesting, and how it relates to this issue. I'm very interested though, so please let me know more :smile:.

**Comment by Virgiel** [comment]
A solution could be to allow the naming of return structure.

For example:

```sql
--! authors() Person*
SELECT name, country FROM authors;

--! clients() Person*
SELECT name, country FROM clients;
```

This would generate a struct like:
```
struct Person {
    name: String,
    country: String
}
```

However, there are some technical challenges. We need to check all queries return the same types associated with the same names, and for ergonomic reasons, we should allow columns to be in any order.

**Comment by Virgiel** [comment]
Then if I want to do some normalization or assertion, I can simply use standards `From` or `TryFrom` traits to map query return struct to mine.

**Comment by Virgiel** [comment]
Another alternative is to be generic over `From`:
```sql
--! authors()*
SELECT name, country FROM authors;
```
Generate:
```pub async fn authors<T: GenericClient, R: From<(String, String)>>(
    client: &T,
) -> Result<impl Stream<Item = Result<R, Error>>, Error> {
    let stmt = client.prepare("SELECT info FROM Author;").await?;
    let stream = client
        .query_raw(&stmt, std::iter::empty::<u32>())
        .await?
        .map(move |res| {
            res.map(|res| {
                let return_value_1: String = res.get(0);
                let return_value_2: String = res.get(1);
                R::from((return_value_1, return_value_2))
            })
        });
    Ok(stream.into_stream())
}
```
Used like that:
```rust

pub struct Person {
    name: String,
    country: String,
}

impl From<(String, String)> for Person {
    fn from((name, country): (String, String)) -> Self {
        Self {
            name: name.trim().to_owned(),
            country: country
                .split('|')
                .map(|s| s.trim())
                .nth(2)
                .unwrap_or_default()
                .to_owned(),
        }
    }
}

pub async fn get_authors<T: GenericClient>(client: &T) -> Result<Vec<Person>, Error> {
    authors(client).await?.try_collect().await
}
```

**Comment by LouisGariepy** [comment]
On track to be solved by #52.
